### PR TITLE
Refactor footer to allow multicolumn and external links

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

The current footer was automatically generated from the structure of the page tree.

Our users did not always manage to structure their footers the way they wanted. They requested in particular that richie should:

- allow organizing footer links in columns or as flat links,
- allow customizing footer links (e.g. by setting the link target),
- allow adding any internal or external link to the footer,
- de-correlate the structure of footer links from the page tree.

## Proposal

This PR proposes to use a static placeholder to define the footer.

- [x] Remove the menu template tag and the related template that were generating a footer from the page tree structure
- [x] Adapt the section plugin so that we can use it to generate unordered lists
- [x] Make the `title` field optional on the section plugin so that our lists don't necessarily have a title... this will also be usefull for highlight sections for which the title was also sometimes annoying,
- [x] Add static placeholder that accepts section and link plugins
- [x] Add a data migration to avoid breaking existing footers
- [x] Add footer as static placeholder in demo site

:information_source:  While working on the section plugin, I also took the opportunity to add missing tests on the existing templates.

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -16,6 +16,14 @@ $ make migrate
 
 ## Unreleased
 
+## 1.13.x to 1.14.x
+
+- Footer links are now defined via a static placeholder (see CHANGELOG). We provide a data
+  migration to make sure that your existing footer of flat links is ported to the new static
```
Comment (jmaupetit) [inline UPGRADE.md]:
```suggestion
  migration to make sure that an existing footer of flat links is ported to the new static
```

File: UPGRADE.md [hunk 2]
```diff
@@ -16,6 +16,14 @@ $ make migrate
...
+  placeholder. Before running the data migration, make sure all your changes on the footer pages
```
Comment (jmaupetit) [inline UPGRADE.md]:
```suggestion
  placeholder. Before running the data migration, make sure all changes on the footer pages
```

File: UPGRADE.md [hunk 3]
```diff
@@ -16,6 +16,14 @@ $ make migrate
 
 ## Unreleased
 
+## 1.13.x to 1.14.x
+
+- Footer links are now defined via a static placeholder (see CHANGELOG). We provide a data
```
Comment (jmaupetit) [inline UPGRADE.md]:
```suggestion
- Footer links are now defined via a static placeholder (see CHANGELOG). Richie provides a data
```

File: UPGRADE.md [hunk 4]
```diff
@@ -16,6 +16,14 @@ $ make migrate
...
+  migration to make sure that your existing footer of flat links is ported to the new static
+  placeholder. Before running the data migration, make sure all your changes on the footer pages
+  (pages that are below the page with a reverse_id of "annex") are published. After running the
+  data migration, check the footer on your site (draft and public) to make sure it is as expected.
```
Comment (jmaupetit) [inline UPGRADE.md]:
```suggestion
  data migration, check the footer (draft and public) to make sure it is as expected.
```

File: src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    Create a footer with the new static placeholder from the existing footer pages that were
+    placed under the "annex" page and displayed in the footer via a `show_menu_below_id`
+    template tag.
+    """
+    Page = apps.get_model("cms", "Page")
+    Title = apps.get_model("cms", "Title")
+
+    # We should import StaticPlaceholder from apps but its `draft` and `public` fields
+    # are custom foreign key field that checks that they are targeting an instance of
+    # # cms.models Placeholder so the code would not work. We can safely assume that the
```
Comment (jmaupetit) [inline src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py]:
```suggestion
    # `cms.models.Placeholder` so the code would not work. We can safely assume that the
```

File: src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    # Placeholder and StaticPlaceholder models is still there when this migration is run
```
Comment (jmaupetit) [inline src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py]:
```suggestion
    # Placeholder and StaticPlaceholder models are still there when this migration is run.
```

File: src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    static_placeholder, was_created = StaticPlaceholder.objects.get_or_create(
+        code="footer"
+    )
+
+    if not was_created:
+        # If the static placeholder was already existing, it means this migration is being
+        # replayed and we better do nothing
```
Comment (jmaupetit) [inline src/richie/apps/courses/migrations/0010_auto_footer_to_static_placeholder.py]:
```suggestion
        # replayed and we better do nothing.
```

File: tests/plugins/section/test_templates_section_cadenced.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            name="Example 2",
+            target=parent,
+        )
+
+        html = context["cms_content_renderer"].render_placeholder(
+            placeholder, context=context, language="en"
+        )
+        self.assertHTMLEqual(
+            html,
+            f"""
```
Comment (jmaupetit) [inline tests/plugins/section/test_templates_section_cadenced.py]:
```suggestion
            """
```

File: tests/plugins/section/test_templates_section_cadenced.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                name="Example 2",
+                target=parent,
+            )
+
+        html = context["cms_content_renderer"].render_placeholder(
+            placeholder, context=context, language="en"
+        )
+        self.assertHTMLEqual(
+            html,
+            f"""
```
Comment (jmaupetit) [inline tests/plugins/section/test_templates_section_cadenced.py]:
```suggestion
            """
```

File: tests/plugins/section/test_templates_section_default.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+            name="Example 2",
+            target=parent,
+        )
+
+        html = context["cms_content_renderer"].render_placeholder(
+            placeholder, context=context, language="en"
+        )
+        self.assertHTMLEqual(
+            html,
+            f"""
```
Comment (jmaupetit) [inline tests/plugins/section/test_templates_section_default.py]:
```suggestion
            """
```

File: tests/plugins/section/test_templates_section_default.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+                name="Example 2",
+                target=parent,
+            )
+
+        html = context["cms_content_renderer"].render_placeholder(
+            placeholder, context=context, language="en"
+        )
+        self.assertHTMLEqual(
+            html,
+            f"""
```
Comment (jmaupetit) [inline tests/plugins/section/test_templates_section_default.py]:
```suggestion
            """
```

File: tests/plugins/section/test_templates_section_default.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            <section class="section">
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply tests/plugins/section/test_templates_section_default.py]: If you want a flat footer, you have only a section containing linkplugins.

If you want a footer with columns of links, the first section wraps everything and the nested sections each contain a list of link plugins.

**Pull request merged**